File: bitdrift-shop/android/app/src/main/java/ai/bitdrift/shop/Screen.kt
package ai.bitdrift.shop

/**
 * Navigation Path - sealed class representing all screens in the app.
 *
 * Step 1: Welcome, Advanced
 * Step 2: Browse, Search
 * Step 3: Featured, Categories
 * Step 4: ProductDetail, Reviews
 * Step 5: Cart, Wishlist
 * Step 6: CheckoutGuest, CheckoutSignIn
 * Step 6: PaymentCard, PaymentApplePay, PaymentPayPal, PaymentAndroidPay
 * Step 7: Confirmation
 */
sealed class Screen(val route: String) {
    // Step 1
    data object Welcome : Screen("welcome")
    data object Advanced : Screen("advanced")

    // Step 2 branches
    data object Browse : Screen("browse")
    data object Search : Screen("search")

    // Step 3 branches
    data object FeaturedProducts : Screen("featured")
    data object Categories : Screen("categories")
    data class CategoryBrowse(val category: String) : Screen("categoryBrowse/$category")

    // Step 4 branches (source = featured|categories, productId passed from browse list)
    data class ProductDetail(val source: String, val productId: String = "") : Screen("productDetail/$source/$productId")
    data class Reviews(val source: String, val productId: String = "") : Screen("reviews/$source/$productId")

    // Step 5 branches (carry productId for cart/wishlist API call)
    data class Cart(val productId: String = "") : Screen("cart/$productId")
    data class Wishlist(val productId: String = "") : Screen("wishlist/$productId")

    // Step 6 branches (carry productId for checkout context)
    data class CheckoutGuest(val productId: String = "") : Screen("checkoutGuest/$productId")
    data class CheckoutSignIn(val productId: String = "") : Screen("checkoutSignIn/$productId")

    // Step 7 branches (carry checkoutSession for payment)
    data class PaymentCard(val checkoutSession: String = "") : Screen("paymentCard/$checkoutSession")
    data class PaymentApplePay(val checkoutSession: String = "") : Screen("paymentApplePay/$checkoutSession")
    data class PaymentPayPal(val checkoutSession: String = "") : Screen("paymentPayPal/$checkoutSession")
    data class PaymentAndroidPay(val checkoutSession: String = "") : Screen("paymentAndroidPay/$checkoutSession")

    // Payment failure (carry paymentMethod + checkoutSession for retry context)
    data class PaymentFailed(val paymentMethod: String = "", val checkoutSession: String = "") : Screen("paymentFailed/$paymentMethod/$checkoutSession")

    // Final step (carry orderId for confirmation)
    data class Confirmation(val orderId: String = "") : Screen("confirmation/$orderId")

    companion object {
        const val PRODUCT_DETAIL_ROUTE = "productDetail/{source}/{productId}"
        const val CATEGORY_BROWSE_ROUTE = "categoryBrowse/{category}"
        const val REVIEWS_ROUTE = "reviews/{source}/{productId}"
        const val CART_ROUTE = "cart/{productId}"
        const val WISHLIST_ROUTE = "wishlist/{productId}"
        const val CHECKOUT_GUEST_ROUTE = "checkoutGuest/{productId}"
        const val CHECKOUT_SIGNIN_ROUTE = "checkoutSignIn/{productId}"
        const val PAYMENT_CARD_ROUTE = "paymentCard/{checkoutSession}"
        const val PAYMENT_APPLEPAY_ROUTE = "paymentApplePay/{checkoutSession}"
        const val PAYMENT_PAYPAL_ROUTE = "paymentPayPal/{checkoutSession}"
        const val PAYMENT_ANDROIDPAY_ROUTE = "paymentAndroidPay/{checkoutSession}"
        const val PAYMENT_FAILED_ROUTE = "paymentFailed/{paymentMethod}/{checkoutSession}"
        const val CONFIRMATION_ROUTE = "confirmation/{orderId}"
    }
}

File: bitdrift-shop/ios/BitdriftShop/Screen.swift
import Foundation

/// Navigation destinations. Route strings are byte-identical to the Android
/// app's so both platforms emit the same `_screen_name` values and share the
/// `bd-shop-*` workflows and dashboards.
///
///     Step 1: Welcome, Advanced
///     Step 2: Browse, Search
///     Step 3: Featured, Categories, CategoryBrowse
///     Step 4: ProductDetail, Reviews
///     Step 5: Cart, Wishlist
///     Step 6: CheckoutGuest, CheckoutSignIn, Payment*, PaymentFailed
///     Step 7: Confirmation
enum Screen: Hashable {
    case welcome
    case advanced

    case browse
    case search

    case featured
    case categories
    case categoryBrowse(category: String)

    case productDetail(source: String, productID: String)
    case reviews(source: String, productID: String)

    case cart(productID: String)
    case wishlist(productID: String)

    case checkoutGuest(productID: String)
    case checkoutSignIn(productID: String)

    case paymentCard(checkoutSession: String)
    case paymentApplePay(checkoutSession: String)
    case paymentPayPal(checkoutSession: String)
    case paymentAndroidPay(checkoutSession: String)

    case paymentFailed(paymentMethod: String, checkoutSession: String)

    case confirmation(orderID: String)

    /// The route string, matching Android's `Screen.route`.
    var route: String {
        switch self {
        case .welcome: return "welcome"
        case .advanced: return "advanced"
        case .browse: return "browse"
        case .search: return "search"
        case .featured: return "featured"
        case .categories: return "categories"
        case .categoryBrowse(let category): return "categoryBrowse/\(esc(category))"
        case .productDetail(let source, let pid): return "productDetail/\(esc(source))/\(esc(pid))"
        case .reviews(let source, let pid): return "reviews/\(esc(source))/\(esc(pid))"
        case .cart(let pid): return "cart/\(esc(pid))"
        case .wishlist(let pid): return "wishlist/\(esc(pid))"
        case .checkoutGuest(let pid): return "checkoutGuest/\(esc(pid))"
        case .checkoutSignIn(let pid): return "checkoutSignIn/\(esc(pid))"
        case .paymentCard(let s): return "paymentCard/\(esc(s))"
        case .paymentApplePay(let s): return "paymentApplePay/\(esc(s))"
        case .paymentPayPal(let s): return "paymentPayPal/\(esc(s))"
        case .paymentAndroidPay(let s): return "paymentAndroidPay/\(esc(s))"
        case .paymentFailed(let m, let s): return "paymentFailed/\(esc(m))/\(esc(s))"
        case .confirmation(let oid): return "confirmation/\(esc(oid))"
        }
    }

    /// The name reported to `Logger.logScreenView`. Matches Android's
    /// `destinationToScreenName()` exactly — these values are what the Sankey
    /// and per-screen crash analytics group on.
    var screenName: String {
        switch self {
        case .welcome: return "Welcome"
        case .advanced: return "Advanced"
        case .browse: return "Browse"
        case .search: return "Search"
        case .featured: return "Featured"
        case .categories: return "Categories"
        case .categoryBrowse: return "CategoryBrowse"
        case .productDetail: return "ProductDetail"
        case .reviews: return "Reviews"
        case .cart: return "Cart"
        case .wishlist: return "Wishlist"
        case .checkoutGuest: return "CheckoutGuest"
        case .checkoutSignIn: return "CheckoutSignIn"
        case .paymentCard: return "PaymentCard"
        case .paymentApplePay: return "PaymentApplePay"
        case .paymentPayPal: return "PaymentPayPal"
        case .paymentAndroidPay: return "PaymentAndroidPay"
        case .paymentFailed: return "PaymentFailed"
        case .confirmation: return "Confirmation"
        }
    }

    /// A path segment may legitimately be empty (e.g. `cart/` when the cart is
    /// opened from the toolbar with no product) or contain a space (category
    /// names like "Home & Garden"), so segments are percent-encoded on the way
    /// in and decoded on the way out — the same job Android's NavType.StringType
    /// argument handling does.
    private func esc(_ value: String) -> String {
        value.addingPercentEncoding(withAllowedCharacters: .alphanumerics) ?? value
    }
}
